storage/reader.go

```diff
@@ -392,6 +525,82 @@ func (r *Reader) Read(p []byte) (int, error) {
 	return n, err
 }
 
+func (r *Reader) readWithGRPC(p []byte) (int, error) {
+	// No stream to read from, either never initiliazed or Close was called.
+	// TODO: should an error be returned here? "reader has been closed"?
+	if r.stream == nil {
+		return 0, io.EOF
+	}
+
+	// The entire object has been read by this reader, return EOF.
+	if r.size != 0 && r.size == r.seen {
+		return 0, io.EOF
+	}
+
+	n := 0
+
+	if len(r.leftovers) > 0 {
+		cp := copy(p, r.leftovers)
+		r.seen += int64(cp)
+		r.leftovers = r.leftovers[cp:]
+
+		// No more room left in the user's buffer or the entire object has been
+		// read.
+		if len(p[cp:]) == 0 || r.size == r.seen {
+			return cp, nil
+		}
+
+		n = cp
+	}
+
+	for len(p[n:]) > 0 {
```

Comment:
Why is the termination condition for this while loop different from the `if()` on line 549?

Reply:
`if len(r.leftovers) > 0 {` just checks to see if the client's internal buffer has any data stashed in it to use for filling the user-provided buffer. `for len(p[n:]) > 0 {` will continue `Recv`'ing until the user-provided buffer has been filled.

Did that answer the question?

Reply:
Your code is probably correct, it just looked odd to me.  Basically I wondered: if there is reason to believe that you should return when `r.size == r.seen` in line 549: do those reasons apply here?  Do you even need that `if()` in line 549?  Since you looked at the code twice, I am fine either way, you know better.

Reply:
I went ahead and just tried removing the conditional and everything still worked. You were right; the `leftovers` will either have nothing (`copy` returns `0`) and the following code just executes no problem with `0`, or it has something and we use it. Since `leftovers` is filled with the left over bytes from the previous `Recv` msg, if the user's buffer has more space, we should try to `Recv` again. 

Thanks Carlos! Definitely cleaned this up.